client/received_blocks_new_map.go

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package client
+
+import (
+	"github.com/cespare/xxhash"
+	"github.com/m3db/m3x/ident"
+	"github.com/m3db/m3x/pool"
+)
+
+func newReceivedBlocksMap(pool pool.BytesPool) *receivedBlocksMap {
+	return _receivedBlocksMapAlloc(_receivedBlocksMapOptions{
+		hash: func(k idAndBlockStart) receivedBlocksMapHash {
+			return receivedBlocksMapHash(xxhash.Sum64(k.id.Bytes())) +
+				receivedBlocksMapHash(k.blockStart)
+		},
+		equals: func(x, y idAndBlockStart) bool {
+			return x.id.Equal(y.id) && x.blockStart == y.blockStart
+		},
+		copy: func(k idAndBlockStart) idAndBlockStart {
+			bytes := k.id.Bytes()
+			keyLen := len(bytes)
+			pooled := pool.Get(keyLen)[:keyLen]
+			copy(pooled, bytes)
+			return idAndBlockStart{
+				id:         ident.BytesID(pooled),
+				blockStart: k.blockStart,
+			}
+		},
+		finalize: func(k idAndBlockStart) {
+			if slice, ok := k.id.(ident.BytesID); ok {
```

Comment:
why is this check needed?

Reply:
Someone might use `SetUnsafe(...)` with NoCopyKey so they have a user defined type as the ID passed in here.  This is safe under that case, also type assertion to a concrete type `BytesID` which is a type alias of `[]byte` is super fast:
https://github.com/teh-cmc/go-internals/blob/master/chapter2_interfaces/README.md#type-assertions

Reply:
Makes sense. 

PS that's a great reference, thanks for the link!